File: Тренировки по алгоритмам 3.0/Дивизион B/Тема 5. Обход графов в глубину/32.py
from collections import defaultdict
import sys

sys.setrecursionlimit(100000)


def input_read():
    n, m = [int(x) for x in input().split()]
    graph = [list() for _ in range(n + 1)]
    for _ in range(m):
        v1, v2 = [int(x) for x in input().split()]
        graph[v1].append(v2)
        graph[v2].append(v1)
    return n, m, graph


def dfs(graph, visited, now, comp_num):
    if visited[now]:
        return
    visited[now] = comp_num
    for v in graph[now]:
        if not visited[v]:
            dfs(graph, visited, v, comp_num)


def components(n, m, graph):
    visited = [0] * (n + 1)
    comp_num = 0
    for v in range(1, n + 1):
        comp_num += 1
        dfs(graph, visited, v, comp_num)
    answer = defaultdict(list)
    for i in range(1, n + 1):
        answer[visited[i]].append(i)
    return answer


def main():
    answer = components(*input_read())
    print(len(answer))
    for component in answer.values():
        print(len(component))
        print(*component)


if __name__ == '__main__':
    main()

File: Тренировки по алгоритмам 3.0/Дивизион B/Тема 5. Обход графов в глубину/32.cpp
#include <algorithm>
#include <cassert>
#include <iostream>
#include <ranges>
#include <string>
#include <unordered_map>
#include <vector>

using namespace std;
namespace rng = std::ranges;
namespace view = rng::views;
using namespace std;

void Dfs(vector<vector<int>>& graph, vector<int>& visited, int now, int comp_num) {
	if (visited[now]) {
		return;
	}
	visited[now] = comp_num;
	for (int v : graph[now]) {
		if (!visited[v]) {
			Dfs(graph, visited, v, comp_num);
		}
	}
}

unordered_map<int, vector<int>> Components(int N, vector<vector<int>>& graph) {
	vector<int> visited(N + 1);
	int comp_num = 0;
	for (int v : view::iota(1, N + 1)) {
		Dfs(graph, visited, v, ++comp_num);
	}
	unordered_map<int, vector<int>> answer;
	for (int i : view::iota(1, N + 1)) {
		answer[visited[i]].push_back(i);
	}
	return answer;
}

int main() {
	int N, M;
	cin >> N >> M;
	vector<vector<int>> graph(N + 1);
	for (auto i : view::iota(0, M)) {
		int v1, v2;
		cin >> v1 >> v2;
		graph[v1].push_back(v2);
		graph[v2].push_back(v1);
	}
	auto answer = Components(N, graph);
	cout << answer.size() << '\n';
	for (const auto& [i, component] : answer) {
		cout << component.size() << '\n';
		for (int v : component) {
			cout << v << ' ';
		}
		cout << '\n';
	}
}
